Add support for multiple Before You Go pages, targeted by Altis audience selector

goldenapples opened the pull request:
* Updates the way that the BYG plugin config is handled to allow for selecting the logic for each BYG experience using the Altis Analytics audience selector.
* Removes hardcoded logic for referers and UTM sources.
* Registers Altis Analytics and HM Asset Loader as dependencies.

goldenapples committed:
Make BYG page post type hierarchical

Defining the post type as hierarchical allow an easy UI for sorting it
in the post list table, which will be useful as an alternative to always
showing the latest chronological post.

goldenapples committed:
Add UI for selecting audience on BYG pages

goldenapples committed:
Register Altis.Analytics as external in webpack config

goldenapples committed:
Define BYG page trioggers by Altis audiences.

Allow for stacking BYG pages by priority and selecting an Altis audience
to target with each one.

goldenapples committed:
Remove "query" and "referrer" trigger logic

goldenapples committed:
Register Asset Loader and AWS Analytics as dependencies

This plugin depends on Asset Loader and Altis Analytics to be installed
and active, but they're not explicitly declared as dependencies. This
adds those to composer.json, so that installing will fail if those
plugins aren't available.

goldenapples committed:
Allow no unit tests

I've removed all the unit tests, so this needs to be able to pass
without any tests.

goldenapples committed:
Fix PHP lint issues and add docblocks

goldenapples committed:
Update composer lockfile

goldenapples commented:
@kadamwhite I'd like to get your thoughts on this approach before I go any further with it. I did a lot of hacking and slashing to get to this point and afterwards realized that I could have handled this with custom override code in the project, if the goal is to keep this plugin completely platform-agnostic. 

But since most of the audience detection logic we'd be implementing here is already supported by Altis Accelerate, I thought it might be nice to just register that as a peer dependency and use it instead of reimplementing all the logic checks required to match utm values or referrers.

kadamwhite commented on .eslintrc.js:
```diff
@@ -276,7 +276,7 @@ module.exports = {
 		'no-this-before-super': 'warn',
 		'no-throw-literal': 'warn',
-		'no-undef': 'error',
+		'no-undef': 'off',
```
Nit: Arguably this would be better as a `warn`, we do generally want to clean those up so disabling it entirely feels wrong

kadamwhite commented on .webpack/webpack.config.dev.js:
```diff
@@ -12,7 +12,10 @@ module.exports = choosePort( 9090 ).then( ( port ) => presets.development( {
 		port,
 	},
-	externals,
+    externals: {
```
Nit, indentation inconsistency

kadamwhite commented on .webpack/webpack.config.dev.js:
```diff
@@ -12,7 +12,10 @@ module.exports = choosePort( 9090 ).then( ( port ) => presets.development( {
-	externals,
+    externals: {
+		...externals,
+		'@altis/analytics': 'Altis.Analytics',
```
Nice

kadamwhite commented on .webpack/webpack.config.prod.js:
```diff
@@ -4,7 +4,10 @@ const { filePath } = helpers;
 module.exports = presets.production( {
     name: 'before-you-go',
-    externals,
+    externals: {
```
Hah, the inconsistency is mine, I see... these files didnt' start out as tabbed indent at all. Wacky.

kadamwhite commented on inc/post-types/byg-page.php:
```diff
@@ -75,22 +80,75 @@ function register_type() {
+}
+
+/**
+ * Add a column for "audience to the post list table for BYG posts.
```
```suggestion
 * Add a column for "audience" to the post list table for BYG posts.
```

kadamwhite reviewed (commented):
I don't think tightly coupling to Accelerate is bad at all, that seems entirely reasonable. Feels a lot simpler than managing referer and utm entirely on our own.

goldenapples committed:
Turn eslint no-undef warning back on

warn is enough here, it doesn't need to be an error tho

goldenapples committed:
Standardize whitespace in webpack config files

tabs >>> spaces

goldenapples committed:
Fix typo in docblock

goldenapples commented on inc/assets.php:
```diff
@@ -130,3 +112,26 @@ function enqueue_frontend_scripts() : void {
+		 *
+		 * @param string[] $dependencies Script handles to load ahead of the frontend bundle.
+		 */
+		apply_filters( 'byg\editor_script_dependencies', [ 'altis-analytics-audience-ui' ] )
```
I'm going to remove this filter as it feels overcomplicated, now that we're relying on Altis Analytics for the bulk of the logic. It's not too cumbersome to alter a registered script and add a new dependency after the fact, so having this filter in place feels unnecessary.

goldenapples committed:
Remove filters around script dependencies

I think these filters might be overcomplicating things, as I'm not sure
where they'd be necessary, and if you did want to add dependencies to
scripts, it's not too complicated to do that using core functionality
anyhow.

goldenapples committed:
Fix HMR for editor plugin

`unregisterPlugin` wasn't being imported, so HMR didn't work here

goldenapples committed:
Merge branch 'main' into add-audience-selector

goldenapples committed:
Update type definitions for page configurations

goldenapples commented:
Finished this PR and did some cleanup throughout. @kadamwhite would you give it another review when you get a chance?
